diff --git a/node_modules/@angular/build/src/builders/application/build-action.js b/node_modules/@angular/build/src/builders/application/build-action.js
index 61d67ff..86a1d40 100755
--- a/node_modules/@angular/build/src/builders/application/build-action.js
+++ b/node_modules/@angular/build/src/builders/application/build-action.js
@@ -94,6 +94,10 @@ async function* runEsBuildBuildAction(action, options) {
             cacheOptions.basePath,
             `${(0, path_1.toPosixPath)(workspaceRoot)}/**/.*/**`,
         ];
+        // Ignore third-party packages under node_modules for Watchpack (unless preserving symlinks for npm link).
+        if (!preserveSymlinks) {
+            ignored.push('**/node_modules/**');
+        }
         // Setup a watcher
         const { createWatcher } = await Promise.resolve().then(() => __importStar(require('../../tools/esbuild/watcher')));
         watcher = createWatcher({
@@ -107,10 +111,7 @@ async function* runEsBuildBuildAction(action, options) {
         // Watch the entire project root if 'NG_BUILD_WATCH_ROOT' environment variable is set
         if (environment_options_1.shouldWatchRoot) {
             if (!preserveSymlinks) {
-                // Ignore all node modules directories to avoid excessive file watchers.
-                // Package changes are handled below by watching manifest and lock files.
-                // NOTE: this is not enable when preserveSymlinks is true as this would break `npm link` usages.
-                ignored.push('**/node_modules/**');
+                // Package changes are handled by watching manifest and lock files.
                 watcher.add(packageWatchFiles
                     .map((file) => node_path_1.default.join(workspaceRoot, file))
                     .filter((file) => (0, node_fs_1.existsSync)(file)));
